plutus-contract/test/Spec/MustSpendAtLeast.hs

```diff
@@ -0,0 +1,127 @@
+import Ledger.Typed.Scripts qualified as Scripts
+import Plutus.Contract as Con
+import Plutus.Contract.Test (assertContractError, assertFailedTransaction, assertValidatedTransactionCount,
+                             checkPredicateOptions, defaultCheckOptions, w1, (.&&.))
+import Plutus.Trace qualified as Trace
+import Plutus.V1.Ledger.Api (Datum (Datum), ScriptContext, Validator, ValidatorHash)
+import Plutus.V1.Ledger.Scripts (ScriptError (EvaluationError))
+import PlutusTx qualified
+import PlutusTx.Prelude qualified as P
+import Prelude hiding (not)
+
+tests :: TestTree
+tests =
+    testGroup "MustSpendAtLeast"
+        [ entireScriptBalance
+        , lessThanScriptBalance
+        , higherThanScriptBalance
+        , phase2Failure
+        ]
+
+scriptBalance :: Integer
+scriptBalance = 25_000_000
+
+contract :: Integer -> Integer -> Contract () Empty ContractError ()
+contract offAmt onAmt = do
+    let lookups1 = Constraints.plutusV1TypedValidatorLookups typedValidator
+        tx1 = Constraints.mustPayToTheScript onAmt (Ada.lovelaceValueOf scriptBalance)
+    ledgerTx1 <- submitTxConstraintsWith lookups1 tx1
+    awaitTxConfirmed $ Tx.getCardanoTxId ledgerTx1
+
+    utxos <- utxosAt scrAddress
+    let lookups2 = Constraints.plutusV1TypedValidatorLookups typedValidator
+            <> Constraints.plutusV1OtherScript validatorScript
+            <> Constraints.unspentOutputs utxos
+        tx2 =
+            Constraints.collectFromTheScript utxos ()
+            <> Constraints.mustIncludeDatum (Datum $ PlutusTx.toBuiltinData onAmt)
+            <> Constraints.mustSpendAtLeast (Ada.lovelaceValueOf offAmt)
+    ledgerTx2 <- submitTxConstraintsWith @UnitTest lookups2 tx2
+    awaitTxConfirmed $ Tx.getCardanoTxId ledgerTx2
+
+trace :: Integer -> Integer -> Trace.EmulatorTrace ()
+trace offAmt onAmt = do
+    void $ Trace.activateContractWallet w1 $ contract offAmt onAmt
+    void $ Trace.waitNSlots 1
+
+entireScriptBalance :: TestTree
+entireScriptBalance = checkPredicateOptions
+    defaultCheckOptions
+    "Successful use of mustSpendAtLeast at script's exact balance"
+    (assertValidatedTransactionCount 2)
+    (void $ trace scriptBalance scriptBalance)
+
+lessThanScriptBalance :: TestTree
+lessThanScriptBalance =
+    let amt = scriptBalance - 1
+    in  checkPredicateOptions
+            defaultCheckOptions
+            "Successful use of mustSpendAtLeast below script's balance"
+            (assertValidatedTransactionCount 2)
+            (void $ trace amt amt )
+
+higherThanScriptBalance :: TestTree
+higherThanScriptBalance =
+    let amt = scriptBalance + 1
+    in  checkPredicateOptions
+            defaultCheckOptions
+            "Fail validation when mustSpendAtLeast is greater than script's balance"
+            (assertContractError (contract amt amt) (Trace.walletInstanceTag w1) (\case { ConstraintResolutionContractError Constraints.OwnPubKeyMissing -> True; _ -> False }) "failed to throw error"
```

Comment:
This expected result doesn't fit with the description "Fail validation when mustSpendAtLeast is greater than script's balance".

If you locked 25ADA in the script, and you want the transaction to spend 26ADA, the 1ADA that is left will be compensated by the wallet provided in the lookups.

Then, you can either change the test case description, or add the pubkey in the lookups, and show that we don't get a failing test (or add a new test case for that)

Reply:
I've split this test out into two, using `ownPaymentPubKeyHash` lookup. They are now both failing due to `ValueContainsLessThanMinAda` by 1 lovelace.

- `higherThanScriptBalanceWithWalletPubkeyLookup` is using the correct own wallet PaymentPubKeyHash. Shouldn't it try and abide by this rule?

- `higherThanScriptBalanceWithoutWalletPubkeyLookup` is using other wallet's PaymentPubKeyHash. Shouldn't it just fail with `OwnPubKeyMissing`?

Reply:
Yes! That's because you can't create an output with less than (I think it's hardcoded to 2) the minAda required for a tx output. You need to make sure that the amount that the provided pubkey needs is higher than that value.

Do `let amt = scriptBalance + 5` instead of `let amt = scriptBalance + 1`

Reply:
Ok, changing to `+ 5` is only lovelace so same error. Chaning to `+ 5_000_000` results in passing contract validation but failing phase-2. Are there two problems here:
1. Constraint passing off-chain validation but failing on-chain
2. `higherThanScriptBalanceWithoutWalletPubkeyLookup` doesn't use w1 pubkeyhash so I'd expect it to throw `OwnPubKeyMissing` error.

Reply:
Bug identified in offchain code, see PLT-665
Have commented out the two affected tests for now.